source/distributed/cell_weights.cc

```diff
@@ -215,6 +219,106 @@ namespace parallel
     // Return the cell weight determined by the function of choice.
     return weighting_function(cell, dof_handler.get_fe(fe_index));
   }
+
+
+
+  template <int dim, int spacedim>
+  std::function<unsigned int(
+    const typename dealii::Triangulation<dim, spacedim>::cell_iterator &cell,
+    const CellStatus                                                    status)>
+  CellWeights<dim, spacedim>::make_cached_weighting_callback(
+    const DoFHandler<dim, spacedim> &dof_handler,
+    const WeightingFunction         &weighting_function)
+  {
+    // build cache for weights
+    std::vector<unsigned int> weight_cache;
+    weight_cache.reserve(dof_handler.get_fe_collection().size());
+
+    const typename DoFHandler<dim, spacedim>::cell_iterator dummy;
+    for (const auto &fe : dof_handler.get_fe_collection())
+      weight_cache.push_back(weighting_function(dummy, fe));
+
+    // create callback function
+    const parallel::TriangulationBase<dim, spacedim> *tria =
+      dynamic_cast<const parallel::TriangulationBase<dim, spacedim> *>(
+        &(dof_handler.get_triangulation()));
+
+    Assert(
+      tria != nullptr,
+      ExcMessage(
+        "parallel::CellWeights requires a parallel::TriangulationBase object."));
+
+    return [&dof_handler, tria, weight_cache](
```

Comment:
You'll want to leave a comment making clear that `weight_cache` is here captured by *copying* the array, not just  by reference.